include/pcp/filter/statistical_outlier_filter.hpp

```diff
@@ -0,0 +1,141 @@
+#ifndef PCP_FILTER_STATISTICAL_OUTLIER_FILTER_HPP
+#define PCP_FILTER_STATISTICAL_OUTLIER_FILTER_HPP
+/**
+ * @file
+ * @ingroup filter
+ */
+
+#include <algorithm>
+#include <iterator>
+#include <map>
+#include <pcp/kdtree/linked_kdtree.hpp>
+#include <vector>
+namespace pcp {
+
+namespace statistical_outlier_filter {
+/**
+ * @brief parameters for the filter
+ * @tparam ScalarType Scalar type for the parameter variables (float,double,etc)
+ */
+template <class ScalarType>
+struct construction_params_t
+{
+    using scalar_type                         = ScalarType;
+    scalar_type std_dev_multiplier_threshold_ = static_cast<float>(1);
+    std::size_t mean_k_                       = 1;
+};
+} // namespace statistical_outlier_filter
+
+/**
+ * @brief The algorithm iterates through the entire input twice:
+ * During the first iteration it will compute the average distance that each point has to its
+ * nearest k neighbors. The value of k can be set using mean_k_ Next, the mean and standard
+ * deviation of all these distances are computed in order to determine a distance threshold. The
+ * distance threshold will be equal to: mean + stddev_mult * stddev. ' The multiplier for the
+ * standard deviation can be set using the std_dev_multiplier_treshold in the params.
+ *
+ * @tparam Element Element type
+ * @tparam ParametersType type of the parameters (double,float)
+ * @tparam PointMap Type satisfying PointMap concept
+ * @tparam KnnMap Type satisfying KnnMap concept
+ */
+template <class Element, class ParametersType, class PointMap, class KnnMap>
+
+class basic_statistical_outlier_filter_t
+{
+  public:
+    using element_type      = Element;
+    using parameters_type   = ParametersType;
+    using parameters_object = statistical_outlier_filter::construction_params_t<parameters_type>;
+    using point_type        = std::invoke_result_t<PointMap, element_type>;
+    template <class ForwardIterator>
+    basic_statistical_outlier_filter_t(
+        ForwardIterator begin,
+        ForwardIterator end,
+        PointMap point_map,
+        KnnMap knn_map,
+        parameters_object params)
+        : point_map_(point_map), knn_map_(knn_map), params_(params)
+    {
+        // size is the number of points in the point cloud
+        auto const& size = std::distance(begin, end);
+        std::vector<element_type> nn_index(params_.mean_k_);
+        std::vector<float> nn_dists(params_.mean_k_);
+        std::transform(
+            begin,
+            end,
+            std::inserter(mean_distances_, mean_distances_.end()),
```

Comment:
This is likely to be very slow, since your unordered_map will need to resize, rehash, resassign elements to buckets many many times, which amounts to adding nodes to lists, which require heap allocation for every node.